Web-components: Dynamic source snippets

bennypowers opened the pull request:
Issue: #11588 

Adds dynamic source to the web-components 'framework' package.

## What I did

Emit the rendered story html (minus lit-html's placeholder comments) to the source channel.

## Questions:

- Can we assume the story is always a lit-html `TemplateResult`? If not then what?
- How do i safely get a reference to the `document`?

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots? **yes**
- [x] Does this need a new example in the kitchen sink apps? **maybe?**
- [x] Does this need an update to the documentation? **probably**

👉 long time lurker first time poster. submitting this single-file PR for discussion and to arouse feedback from our good maintainers.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

bennypowers committed:
feat(web-components): dynamic source

Adds dynamic source to the web-components 'framework' package.

Westbrook commented:
In https://github.com/storybookjs/storybook/blob/next/app/web-components/src/client/preview/render.ts#L24 you can see that the `web-components` build accepts a few other story html shapes: TemplateResult, String, and Node. Would this need to be taken into account, or is the renderer HTML here a little less effected by the real source?

bennypowers commented:
I dunno, might have to 🦆 type the return value of `storyFn`. I'd like to get some tests on this

Westbrook commented:
Question re: decorators for the Storybook team...

Is is possible to resolve the root/origin story, so that the source here comes from that no matter the order in which the decorator is applied? Right now, it has to go first, or it will render the decorator code _only_, which isn't too bad, but would be nice not to require.

bennypowers committed:
test(web-components): dynamic stories

bennypowers commented:
Tried to test this locally but got 

```
 FAIL  addons/docs/src/frameworks/web-components/sourceDecorator.test.ts
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/en/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /Users/bennyp/Developer/storybook/scripts/jest.init.ts:8
    import "core-js/modules/es.array.includes.js";
    ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)

Test Suites: 2 failed, 2 total
Tests:       0 total
Snapshots:   0 total
Time:        2.741 s
Ran all test suites matching /addons\/docs\/src\/frameworks\/web-components\//i
```

🤷‍♂️

bennypowers committed:
style(web-components): refactor to TS

Split config functions into separate modules

Add type and language to docs.source

bennypowers committed:
fix(web-components): types

bennypowers committed:
docs(web-components): add web-components to Dynamic Source support table

bennypowers committed:
docs(web-components): kitchen sink dynamic source example

shilman committed:
Merge branch 'next' into pr/15337

shilman committed:
Fix jest with lit-html ESM

bennypowers committed:
fix(addons): webcomponents transformSource

bennypowers committed:
style(addons): remove null check

shilman commented:
@bennypowers doesn't seem to be working for me. what am i missing?

<img width="1224" alt="https___deploy-preview-15337--storybookjs_netlify_app" src="https://user-images.githubusercontent.com/488689/124941487-6827d580-e03d-11eb-9f86-f3f0dc0a2011.png">

bennypowers commented:
@shilman I think this is intended behaviour.

https://github.com/storybookjs/storybook/blob/209cdac1b76d263056561e23729f62c949b42532/examples/web-components-kitchen-sink/src/components/sb-button.stories.ts#L20-L26

TL;DR: only changes to `primary` will affect the source output because that's how the example _story_ is written.

if you change those lines to the following (and similarly added `background-color` attribute to the descriptor for that property), then I think you'd see a source update.

```js
const Template: Story<SbButton> = ({ primary, backgroundColor, size, label }) =>
  html`<sb-button
    ?primary="${primary}"
    size="${size}"
    label="${label}"
    background-color="${backgroundColor}"
  ></sb-button>`;
  ```

I've made those changes and updated this PR, so once CI is done you should see it reflected on the page

NOTE: there appears to be a problem with the argTypes derived from the manifest for the `size` attribute, which has a type that's a union of strings. I believe that is unrelated to this PR, though

shilman commented:
@bennypowers i don't know enough about web-components to know what is expected behavior, but it looks off to me. naively, i would expect it to produce source strings like:

```
<sb-button
    ?primary=true
    size="large"
    label="hahaha"
    background-color="#ff0000"
  ></sb-button>
```

cc @gaetanmaisse

bennypowers commented:
>  i would expect it to produce source strings like:
> 
> ```
> <sb-button
>     ?primary=true
>     size="large"
>     label="hahaha"
>     background-color="#ff0000"
>   ></sb-button>
> ```
> 
> cc @gaetanmaisse

The purpose here is to output html that someone could copy and paste into any document, rather than, say, a lit-html template javascript source.

> i don't know enough about web-components to know what is expected behavior

web components are one thing, and lit-html is another
the story originally applied DOM properties to the component, which wouldn't affect the output html unless the component was programmed to reflect those properties to attributes, which happens to be the case here with the `primary` property, however that was (originally) the only property assigned in the story as an attr.

bennypowers committed:
docs(web-components): set attributes in example story

bennypowers commented:
> isn't it strange, though, to see a picture of a large, red button that says "hahaha", and paste a snippet that results in a small, blue button that says "Button"

I'm not sure I follow...
Given

<img width="1017" alt="Screen Shot 2021-07-08 at 19 35 30" src="https://user-images.githubusercontent.com/1466420/124959230-ad3f0e00-e023-11eb-874f-06e1d9f6350c.png">

If I paste the snippet into the preview html

<img width="461" alt="Screen Shot 2021-07-08 at 19 36 10" src="https://user-images.githubusercontent.com/1466420/124959347-c8aa1900-e023-11eb-9afc-f893129ee140.png">

I get

<img width="292" alt="Screen Shot 2021-07-08 at 19 36 14" src="https://user-images.githubusercontent.com/1466420/124959358-cba50980-e023-11eb-8b08-7165daeb9b0f.png">

shilman commented:
i was referring to the behavior before your last set of changes, which you described as expected. are you saying that people should only use lit-html attributes, not properties, if they want their snippets to reflect what's being rendered in the story? i could be ok with that, i just want to understand the limitations of this approach, since changing them from properties to attributes feels like a workaround.

bennypowers commented:
> i was referring to the behavior before your last set of changes, which you described as expected. are you saying that people should only use lit-html attributes, not properties, if they want their snippets to reflect what's being rendered in the story? i could be ok with that, i just want to understand the limitations of this approach, since changing them from properties to attributes feels like a workaround.

Yes, I'm saying that this should be the "best practice" recommendation:

```js
const Template: Story<SbButton> = ({ primary, backgroundColor, size, label }) =>
  html`<sb-button
    ?primary="${primary}"
    size="${ifDefined(size)}"
    label="${ifDefined(label)}"
    background-color="${ifDefined(backgroundColor)}"
  ></sb-button>`;
```

Which is in line with the general community standards - accept attributes if you can, properties for complex data in lieu of children. Components that follow those recommendations are more easily styled in `<noscript>` for example, and can pass between frameworks more easily.

Certainly here, where the goal is a copy-able snippet, content is king, so attrs should be the norm.

Westbrook commented:
When I copied this approach locally, I saw what I expected from the output of the snippet, HTML that I can paste into another page and it just works. I’d prefer if it were `context.originalStoryFn()` so as to omit any decorators, but that might just be me. 

It is possible that I missed the HTML updating with the Controls, but it is HTML that I’d want here so as not to require my consumers to work in my context, be that lit-html, JS string, etc. Assuming that’s true, I can’t fully check while remote from the beach this week, I’m a huge fan of this addition!

shilman reviewed (approved):
Thanks for the clarification @bennypowers @Westbrook. I agree it's a wonderful addition and I'm really excited about getting it merged!

I'd love it if @gaetanmaisse or @brion-fuller can take a look.

Westbrook commented on addons/docs/src/frameworks/web-components/sourceDecorator.ts:
```diff
@@ -0,0 +1,37 @@
+}
+
+export function sourceDecorator(storyFn: StoryFn, context: StoryContext) {
+  const story = storyFn();
```
How do others feel about `context.originalStoryFn` here?

shilman replied:
Here are my thoughts on the matter:
https://github.com/storybookjs/storybook/pull/14652#issuecomment-846549496

Since this is all brand new in web-components, we can make it opt-out OR opt-in, depending on how you guys feel. I believe it's safer to include the decorators by default and allow users to opt-out of the decorators. In the React world, it's quite common for stories to rely on the decorators to provide context to render properly. But I'll defer to your collective judgment for what's best in the web-components world.

bennypowers replied:
@Westbrook can you elaborate on the case, both ways?

Westbrook replied:
Yeah, I’m less worried as long as it’s configurable. However, for my library I use a very Storybook custom decorator to control theme color/scale/direction in a way that you wouldn’t want to require a copy/paster to clean up after the fact. `storyFn()` leaves all of the decorator code whereas without it you copy just the component under demo.

shilman replied:
@bennypowers @Westbrook WDYT about adding configuration in this PR. i'm afraid if it doesn't get done now it will slip through the cracks.

bennypowers replied:
I'm happy to put this in i'm just unfamiliar with the relevant apis.

is it just
```ts
const story = context.stripDecorators ? context.originalStoryFn() : storyFn(): 
```

??

shilman replied:
Let's make it consistent with React:

```js
const story = context?.parameters.docs?.source?.excludeDecorators
    ? context.originalStoryFn(context.args) 
    : storyFn();
```

bennypowers replied:
Added with unit tests, thanks

brion-fuller commented on addons/docs/src/frameworks/web-components/sourceDecorator.ts:
```diff
@@ -0,0 +1,37 @@
+  if (!skipSourceRender(context)) {
+    const container = window.document.createElement('div');
+    render(story, container);
+    const source = applyTransformSource(container.innerHTML.replace(/<!---->/g, ''), context);
```
We should think about how to handle properties as for some cases the source wouldn't really show the state of the component.  This is a little different from frameworks as the difference between attributes and properties are abstracted away.  My initial idea is to a script tag that populates properties.

bennypowers replied:
That sounds like a good idea we could put something together with hast.

I'm wondering if we could leave that off for a later pr though as the current code already provides a lot of value

bennypowers committed:
fix(addons): webcomponents excludeDecorators
